### PR TITLE
gh-119786: Move garbage collection doc from devguide to InternalDocs

### PR DESCRIPTION
**Pull request by iritkatriel**
Co-Authored-By: Adam Turner <9087854+aa-turner@users.noreply.github.com>
Co-Authored-By: Carol Willing <carolcode@willingconsulting.com>
Co-Authored-By: Ezio Melotti <ezio.melotti@gmail.com>
Co-Authored-By: Hugo van Kemenade <hugovk@users.noreply.github.com>
Co-Authored-By: Itamar Ostricher <itamarost@gmail.com>
Co-Authored-By: Jesús Cea <jcea@jcea.es>
Co-Authored-By: Joannah Nanjekye <33177550+nanjekyejoannah@users.noreply.github.com>
Co-Authored-By: Ned Batchelder <ned@nedbatchelder.com>
Co-Authored-By: Pablo Galindo Salgado <Pablogsal@gmail.com>
Co-Authored-By: Pamela Fox <pamela.fox@gmail.com>
Co-Authored-By: Sam Gross <colesbury@gmail.com>
Co-Authored-By: Stefan Pochmann <609905+pochmann@users.noreply.github.com>
Co-Authored-By: T. Wouters <thomas@python.org>
Co-Authored-By: q-ata <24601033+q-ata@users.noreply.github.com>
Co-Authored-By: slateny <46876382+slateny@users.noreply.github.com>
Co-Authored-By: Борис Верховский <boris.verk@gmail.com>

* Issue: gh-119786

### CODE REVIEW

**Comment by JacobCoffee** [comment]
Rendered doc: https://github.com/iritkatriel/cpython/blob/gc-doc/InternalDocs/garbage_collector.md
Old guide: https://devguide.python.org/internals/garbage-collector/

File: InternalDocs/garbage_collector.md [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+object. Such a place could be another object, or a global (or static) C variable, or
+a local variable in some C function. When an object’s reference count becomes zero,
+the object is deallocated. If it contains references to other objects, their
+reference counts are decremented. Those other objects may be deallocated in turn, if
+this decrement makes their reference count become zero, and so on. The reference
+count field can be examined using the ``sys.getrefcount`` function (notice that the
+value returned by this function is always 1 more as the function also has a reference
+to the object when called):
+
+```
```
Comment (JacobCoffee) [inline InternalDocs/garbage_collector.md]: Do we get fancy colors if we specify here?
```suggestion
```pycon
```

File: InternalDocs/garbage_collector.md [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+    3
+    >>> del y
+    >>> sys.getrefcount(x)
+    2
+```
+
+The main problem with the reference counting scheme is that it does not handle reference
+cycles. For instance, consider this code:
+
+```
```
Comment (JacobCoffee) [inline InternalDocs/garbage_collector.md]:
```suggestion
```pycon
```

File: InternalDocs/garbage_collector.md [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+    >>> container = []
+    >>> container.append(container)
+    >>> sys.getrefcount(container)
+    3
+    >>> del container
+```
+
+In this example, ``container`` holds a reference to itself, so even when we remove
```
Comment (JacobCoffee) [inline InternalDocs/garbage_collector.md:42]:
```suggestion
In this example, `container` holds a reference to itself, so even when we remove
```
sphinx->md

File: InternalDocs/garbage_collector.md [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+- When representing data structures like graphs, it is very typical for them to
+  have internal links to themselves.
+
+To correctly dispose of these objects once they become unreachable, they need
+to be identified first.  To understand how the algorithm works, let’s take
+the case of a circular linked list which has one link referenced by a
+variable ``A``, and one self-referencing object which is completely
+unreachable:
+
+```
```
Comment (JacobCoffee) [inline InternalDocs/garbage_collector.md]:
```suggestion
```pycon
```

File: InternalDocs/garbage_collector.md [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+
+The objective is to identify all the unreachable objects.  The collector does
+this by identifying reachable objects; the remaining objects must be
+unreachable.  The first step is to identify all of the "to scan" objects that
+are **directly** reachable from outside the set of candidate objects.  These
+objects have a refcount larger than the number of incoming references from
+within the candidate set.
+
+Every object that supports garbage collection will have an extra reference
+count field initialized to the reference count (``gc_ref`` in the figures)
```
Comment (JacobCoffee) [inline InternalDocs/garbage_collector.md:237]:
```suggestion
count field initialized to the reference count (`gc_ref` in the figures)
```

File: InternalDocs/garbage_collector.md [hunk 6]
```diff
@@ -0,0 +1,596 @@
...
+of that object when the algorithm starts. This is because the algorithm needs
+to modify the reference count to do the computations and in this way the
+interpreter will not modify the real reference count field.
+
+![gc-image1](images/python-cyclic-gc-1-new-page.png)
+
+The GC then iterates over all containers in the first list and decrements by one the
+``gc_ref`` field of any other object that container is referencing.  Doing
```
Comment (JacobCoffee) [inline InternalDocs/garbage_collector.md]:
```suggestion
`gc_ref` field of any other object that container is referencing.  Doing
```

Comment (JacobCoffee) [reply InternalDocs/garbage_collector.md:245]: may be a lot less noise to do a find and replace on \``

File: InternalDocs/garbage_collector.md [hunk 7]
```diff
@@ -0,0 +1,596 @@
...
+    >>> import gc
+    >>> gc.get_threshold()
+    (700, 10, 10)
+```
+
+The content of these generations can be examined using the
+``gc.get_objects(generation=NUM)`` function and collections can be triggered
+specifically in a generation by calling ``gc.collect(generation=NUM)``.
+
+```
```
Comment (JacobCoffee) [inline InternalDocs/garbage_collector.md]:
```suggestion
```pycon
```

File: InternalDocs/garbage_collector.md [hunk 8]
```diff
@@ -0,0 +1,596 @@
...
+collection starts. Initially only generation 0 is examined. If generation 0 has
+been examined more than ``threshold_1`` times since generation 1 has been
+examined, then generation 1 is examined as well. With generation 2,
+things are a bit more complicated; see
+[Collecting the oldest generation](#Collecting-the-oldest-generation) for
+more information. These thresholds can be examined using the
+[`gc.get_threshold`](https://docs.python.org/3/library/gc.html#gc.get_threshold)
+function:
+
+```
```
Comment (JacobCoffee) [inline InternalDocs/garbage_collector.md]:
```suggestion
```pycon
```

File: InternalDocs/garbage_collector.md [hunk 9]
```diff
@@ -0,0 +1,596 @@
...
+  Dictionaries are untracked when created. If a tracked item is inserted into a
+  dictionary (either as a key or value), the dictionary becomes tracked. During a
+  full garbage collection (all generations), the collector will untrack any dictionaries
+  whose contents are not tracked.
+
+The garbage collector module provides the Python function ``is_tracked(obj)``, which returns
+the current tracking status of the object. Subsequent garbage collections may change the
+tracking status of the object.
+
+```
```
Comment (JacobCoffee) [inline InternalDocs/garbage_collector.md]:
```suggestion
```pycon
```

File: InternalDocs/garbage_collector.md (hunk 3 above)
Comment (AA-Turner) [reply InternalDocs/garbage_collector.md:42]: Probably not worth doing these replacements? ``` `` ``` works fine in markdown for literal syntax (you can have an arbitrary number of `` ` ``).

File: InternalDocs/garbage_collector.md [hunk 10]
```diff
@@ -0,0 +1,596 @@
...
+========
+
+The main garbage collection algorithm used by CPython is reference counting. The basic idea is
+that CPython counts how many different places there are that have a reference to an
+object. Such a place could be another object, or a global (or static) C variable, or
+a local variable in some C function. When an object’s reference count becomes zero,
+the object is deallocated. If it contains references to other objects, their
+reference counts are decremented. Those other objects may be deallocated in turn, if
+this decrement makes their reference count become zero, and so on. The reference
+count field can be examined using the ``sys.getrefcount`` function (notice that the
```
Comment (AA-Turner) [inline InternalDocs/garbage_collector.md]:
```suggestion
count field can be examined using the ``sys.getrefcount()`` function (notice that the
```

File: InternalDocs/garbage_collector.md [hunk 11]
```diff
@@ -0,0 +1,596 @@
...
+In order to decide when to run, the collector keeps track of the number of object
+allocations and deallocations since the last collection. When the number of
+allocations minus the number of deallocations exceeds ``threshold_0``,
+collection starts. Initially only generation 0 is examined. If generation 0 has
+been examined more than ``threshold_1`` times since generation 1 has been
+examined, then generation 1 is examined as well. With generation 2,
+things are a bit more complicated; see
+[Collecting the oldest generation](#Collecting-the-oldest-generation) for
+more information. These thresholds can be examined using the
+[`gc.get_threshold`](https://docs.python.org/3/library/gc.html#gc.get_threshold)
```
Comment (AA-Turner) [inline InternalDocs/garbage_collector.md]:
```suggestion
[`gc.get_threshold()`](https://docs.python.org/3/library/gc.html#gc.get_threshold)
```

**Comment by JacobCoffee** [comment]
TIL `pycon`

**Comment by AA-Turner** [comment]
There's also `pytb`. Though unhelpfully Sphinx uses Pygments and GitHub uses Linguist:

https://github.com/github-linguist/linguist/blob/f164d13fa618023ecf2d8f2ed9a6ce5fae731346/lib/linguist/languages.yml#L5727-L5786

File: InternalDocs/garbage_collector.md (hunk 4 above)
Comment (iritkatriel) [reply InternalDocs/garbage_collector.md]: What does this do? I don't notice a difference.

Comment (JacobCoffee) [reply InternalDocs/garbage_collector.md]: its **supposed** to do syntax highlighting, did it not work here? 😬

**Review by JacobCoffee — APPROVED** [review]

**Review by pablogsal — APPROVED** [review]
Thanks @iritkatriel :heart:

**Review by willingc — APPROVED** [review]
Small nit re: markdown style. It's more common to use single backticks to surround words to render as code. Typically, one would use double backticks only if there was a backtick in the code phrase.

**Pull request merged**